### PR TITLE
EZP-30760: Fixed broken link after historical parent location removed

### PR DESCRIPTION
**Pull request by michal-myszka**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30760](https://jira.ez.no/browse/EZP-30760)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `7.5`, `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

This PR resolves problem with broken links created by scenario described in https://jira.ez.no/browse/EZP-30760 
and save the historical URL alias, even if historical location has been deleted (for example, alias /deleted-location/moved-somewhere-else-content will works fine)

There is included also nop-actions clearing for ezplatform:urls:regenerate-aliases (clear all unused nop-links (without parents))

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php [hunk 1]
```diff
@@ -1443,6 +1443,89 @@ public function repairBrokenUrlAliasesForLocation(int $locationId)
...
+                )
+            )
+            ->setParameter('actionType', 'nop');
+
+        return $qb->execute();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php]:
```suggestion
    /**
     * @throws \Doctrine\DBAL\DBALException
     */
```

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php [hunk 2]
```diff
@@ -1443,6 +1443,89 @@ public function repairBrokenUrlAliasesForLocation(int $locationId)
...
+        $qb
+            ->select($platform->getCountExpression('id'))
+            ->from($this->table)
+            ->where(
+                $qb->expr()->eq(
+                    'parent',
+                    $qb->createPositionalParameter($parentId, ParameterType::INTEGER)
+                )
+            )->andWhere(
+                $qb->expr()->eq(
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php]: Please extract `expr()` call to a variable, given there's more than one.

File: eZ/Publish/Core/Persistence/Legacy/Tests/Content/UrlAlias/UrlAliasHandlerTest.php [hunk 3]
```diff
@@ -2575,6 +2576,20 @@ public function testLocationDeleted()
...
+        self::assertEquals(
+            $countAfterFirstDeleting - 1,
+            $this->countRows()
+        );
+        self::assertCount(
+            1,
+            $handler->listURLAliasesForLocation(10)
+        );
+
+        $handler->lookup('to-delete-folder/article-to-move-relative-alias');
```
[the remark this answers is not in the document]
Comment (michal-myszka) [reply eZ/Publish/Core/Persistence/Legacy/Tests/Content/UrlAlias/UrlAliasHandlerTest.php]: Added to `eZ/Publish/API/Repository/Tests/LocationServiceTest.php`

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php [hunk 4]
```diff
@@ -1443,6 +1443,81 @@ public function repairBrokenUrlAliasesForLocation(int $locationId)
...
+        $expressionBuilder = $queryBuilder->expr();
+
+        $selectQueryBuilder
+            ->select('al1.id AS inner_id')
+            ->from($this->table, 'al1')
+            ->leftJoin(
+                'al1',
+                $this->table,
+                'al2',
+                $expressionBuilder->eq('inner_id', 'al2.parent')
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php]:
```suggestion
                $expressionBuilder->eq('al1.id.', 'al2.parent')
```
But please don't use `al1` and `al2`. It's not very meaningful. One of them is parent, the other is child, so there's a lot of possibilities (e.g. `u_parent` and `u` respectively) for naming.

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php [hunk 5]
```diff
@@ -1443,6 +1443,81 @@ public function repairBrokenUrlAliasesForLocation(int $locationId)
...
+            )
+            ->where(
+                $expressionBuilder->eq(
+                    'al1.action_type',
+                    ':actionType'
+                )
+            )
+            ->groupBy('inner_id')
+            ->having($platform->getCountExpression('al2.id').' = 0');
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php]:
```suggestion
            ->having($expressionBuilder->eq($platform->getCountExpression('al2.id'), 0);
```

**Comment by pspanja** [comment]
Just for info, this is described in https://jira.ez.no/browse/EZP-20773 and is related to another URL alias problem described in https://jira.ez.no/browse/EZP-20775.

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php [hunk 6]
```diff
@@ -1443,6 +1443,83 @@ public function repairBrokenUrlAliasesForLocation(int $locationId)
...
+            )
+            ->groupBy('u_parent.id')
+            ->having(
+                $expressionBuilder->eq($platform->getCountExpression('u.id'),0)
+            );
+
+        $wrapperQueryBuilder
+            ->select('inner_id')
+            ->from(
+                sprintf('(%s) wrapper', $selectQueryBuilder)
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php:1482]: hmm, wrapper is an alias so it should be set as 2nd argument.
Would that work:
```suggestion
                $selectQueryBuilder->getSQL(), 'wrapper'
```
?

Comment (michal-myszka) [reply eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php]: Almost, below code works:
`sprintf('(%s)', $selectQueryBuilder), 'wrapper'`

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php [hunk 7]
```diff
@@ -1443,6 +1443,82 @@ public function repairBrokenUrlAliasesForLocation(int $locationId)
...
+    /**
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function getAllChildrenAliases(int $parentId): array
+    {
+        $queryBuilder = $this->connection->createQueryBuilder();
+        $expressionBuilder = $queryBuilder->expr();
+
+        $queryBuilder
+            ->select('*')
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php:1505]: I'd avoid `*` due to performance.
Return the required set of columns
```suggestion
            ->select('id', 'parent', 'text_md5')
```
(maybe id is not even needed).

**Comment by lserwatka** [comment]
@alongosz could you merge it?

**Comment by alongosz** [comment]
Merged up and horizontally as ezsystems/ezplatform-kernel@6ae97c9 into `ezplatform-kernel:1.0` and `ezplatform-kernel:master`.

**Pull request merged**